src/number_theory/bernoulli.lean

```diff
@@ -284,37 +284,35 @@ begin
   let f : ℕ → ℕ → ℚ := λ (a : ℕ), λ (b : ℕ),
   bernoulli a / ↑a! * (coeff ℚ (b + 1)) (exp ℚ ^ n),
   rw [finset.nat.sum_antidiagonal_eq_sum_range_succ f],
-  refine finset.sum_congr _ _,
-  { refl, },
-  { intros m h,
-    simp only [f, exp_pow_eq_rescale_exp, rescale, one_div, coeff_mk, ring_hom.coe_mk, coeff_exp,
-      ring_hom.id_apply, nat.cast_mul, rat.algebra_map_rat_rat],
-    simp only [finset.mem_range] at h,
-    rw [nat.choose_eq_factorial_div_factorial (le_of_lt h)],
-    have hq : (q.succ! : ℚ) ≠ 0,
-    { by_contradiction g,
-      simp only [not_not, nat.cast_succ, nat.factorial_succ, nat.cast_eq_zero,
-        nat.cast_mul, mul_eq_zero] at g,
-      cases g,
-      { exact nat.cast_add_one_ne_zero q g,},
-      { exact nat.factorial_ne_zero q g, }, },
-    rw [eq_comm],
-    rw [div_eq_iff hq],
-    have hqm1 : q - m + 1 = q + 1 - m,
-    { rw [nat.sub_add_comm],
-      apply nat.le_of_lt_succ,
-      exact h, },
-    rw [nat.succ_eq_add_one, mul_assoc _ _ ↑(q.succ!),
-      mul_comm _ ↑(q.succ!), ←mul_assoc, div_mul_eq_mul_div,
-      mul_comm (n ^ (q - m + 1) : ℚ) _, ←mul_assoc _ _ (n ^ (q - m + 1) : ℚ), ←one_div, mul_one_div,
-      div_div_eq_div_mul, hqm1],
-    rw [nat.cast_dvd],
-    { rw [nat.cast_mul],
-      ring, },
-    { simp only [nat.factorial_mul_factorial_dvd_factorial (le_of_lt h)], },
-    { simp only [ne.def, nat.cast_eq_zero, mul_eq_zero, nat.factorial_ne_zero, not_false_iff,
-        or_self], },
-  },
+  refine finset.sum_congr rfl _,
+  intros m h,
+  simp only [f, exp_pow_eq_rescale_exp, rescale, one_div, coeff_mk, ring_hom.coe_mk, coeff_exp,
+    ring_hom.id_apply, nat.cast_mul, rat.algebra_map_rat_rat],
+  simp only [finset.mem_range] at h,
+  rw [nat.choose_eq_factorial_div_factorial (le_of_lt h)],
+  have hq : (q.succ! : ℚ) ≠ 0,
+  { by_contradiction g,
+    simp only [not_not, nat.cast_succ, nat.factorial_succ, nat.cast_eq_zero,
+      nat.cast_mul, mul_eq_zero] at g,
+    cases g,
+    { exact nat.cast_add_one_ne_zero q g,},
+    { exact nat.factorial_ne_zero q g, }, },
```

Comment:
This looks like overkill to me - can't you just use `nat.factorial_ne_zero` on `q.succ`?

Reply:
That works, cool, `norm_cast` seems to be useful.. .